### ISSUE TITLE
Sync Error - upload archive: upload send: EOF

### ISSUE THREAD

**Issue by jessebye** [devspace-cloud/devspace#892]
**What happened?**  
`devspace sync` fails with `Sync Error on /path/to/files: upstream: apply changes: apply creates: upload archive: upload send: EOF`. 

Log showing the upload failures and then the Sync Error:
```
[info]   Start syncing
[done] √ Sync started on /Users/jessebye/repos/pronode <-> . (Pod: need-plantations/pronode-579d65f7dd-wz74v)
[info]   Upstream - Handling 1 removes
[info]   Upstream - Upload 1604 create changes (size 10177223)
[info]   Upstream - Retry upload because of error: EOF
[info]   Upstream - Upload 1604 create changes (size 10177223)
[info]   Upstream - Retry upload because of error: EOF
[info]   Upstream - Upload 1604 create changes (size 10177223)
[info]   Upstream - Retry upload because of error: EOF
[info]   Upstream - Upload 1604 create changes (size 10177223)
[info]   Upstream - Retry upload because of error: EOF
[info]   Upstream - Upload 1604 create changes (size 10177223)
[error]  Sync Error on /Users/jessebye/repos/pronode: upstream: apply changes: apply creates: upload archive: upload send: EOF
```

**What did you expect to happen instead?**  
Local changes should have been uploaded with no error.

**How can we reproduce the bug?** (as minimally and precisely as possible)  
Using devspace version 4.4.0
Try to use `devspace sync`
Observe the issue.

**Local Environment:**  
- Operating System: mac (Catalina)
- Deployment method: kubectl apply
- Devspace version: 4.4.0

**Kubernetes Cluster:**  
- Cloud Provider: aws
- Kubernetes Version: `version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-c0eccc", GitCommit:"c0eccca51d7500bb03b2f163dd8d534ffeb2f7a2", GitTreeState:"clean", BuildDate:"2019-12-22T23:14:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}`

**Anything else we need to know?**  
n/a

/kind bug

**Comment by FabianKramm** [comment]
@jessebye thanks for opening this issue! The EOF error usually occurs when devspace tries to upload a file that changes the size while it is being tared. Is a program operating on these files while you try to upload them? Could you try to sync these files without any potential program modifying these files? Usually this can happen (although it shouldn't happen often), however devspace will just try to reupload and eventually succeed in doing so.

**Comment by jessebye** [comment]
@FabianKramm thanks for the context on why this usually happens! I'll try closing out anything that could be modifying files and see if it keeps happening.

**Comment by abh** [comment]
I keep getting this error, too. I don't believe I have any local files open... It'd be helpful if devspace at least would tell which file it's erring out on.

**Comment by abh** [comment]
One situation I think I can reproduce is if the target directory has files owned by root (and the container runs as another user). It'd be really helpful if the sync log would tell why the receiving process quit / errored out.  I fixed all the file permissions (best I can tell), but I still get

```
 info    Upstream - Retry upload because of error: EOF                                                                                                      2020-04-10T02:52:52-07:00
 info    Upstream - Upload 808 create changes (size 42367095)                                                                                               2020-04-10T02:52:52-07:00
 info    Upstream - Retry upload because of error: EOF                                                                                                      2020-04-10T02:52:57-07:00
 info    Upstream - Upload 808 create changes (size 42367095)                                                                                               2020-04-10T02:52:57-07:00
```

... endlessly.

**Comment by LukasGentele** [comment]
@abh Thanks for trying to provide more details on this. We're really eager to fix this but cannot reproduce it so far. Is it possible for you to share a project with us including devspace.yaml, so we can get a better chance to reproduce this issue? If it is private code, happy to send you an NDA to ensure confidentiality you share code in a private git repository or fork. 

If you can create a reproducible example project, the following information would also help to reproduce this:
- `devspace --version`
- Kubernetes provider: gke, eks, aks, minikube, docker kubernetes
- `kubectl version`

**Comment by abh** [comment]
Done -- I think the thing that makes the error message not go back to the client is when the files to upload exceed a certain size; this reproduces it for me. My cluster is remote, so if you have a local cluster you might need a bigger or more test.bin files.

https://github.com/abh/devspace-eof-test

**Comment by abh** [comment]
I ran the test on digital ocean kubernetes; using DevSpace 4.11.1.

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.0-alpha.1.250+822a2de262372d", GitCommit:"822a2de262372db4fb11aa4d69389c0cdbd869aa", GitTreeState:"clean", BuildDate:"2019-12-31T07:20:47Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.6", GitCommit:"72c30166b2105cd7d3350f2c28a219e6abcd79eb", GitTreeState:"clean", BuildDate:"2020-01-18T23:23:21Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by LukasGentele** [comment]
Thanks @abh. I just ran it and I can reproduce it. I will try to investigate the cause of this now. May take some time.

**Comment by LukasGentele** [comment]
TL;DR: to fix this for you: Just add this to your sync section within `devspace.yaml`:
```
excludePaths:
- .git/
```
Please confirm that this actually works for you as well.

To provide some background on this issue:

As @FabianKramm was already pointing out, the issue arises when files are being transferred by the sync and then they change. In this case the issue is the .git folder. While the sync is uploading the file, it changes locally due to processes in your IDE I assume (at least that was the case for me).

The reason why we couldn't reproduce this for the longest time is that our team generally has the best practice to add `.git/` to `.dockerignore` and `devspace init` automatically adds patterns from `.dockerignore` to the excludePaths within the config. You can actually see that in our quickstart projects. Additionally, this error only occurs if the sync actually needs to sync

I will change the behavior of `devspace init` to always add `.git/` to `excludePaths`, so this will become a default for new users. I will also add a hint to the docs to make sure users realize this issue.

Another thing I will add to the docs is that there is actually a very good way to get more details about sync errors:
```bash
devspace deploy                                                         # must deploy first to be able to sync
devspace sync --verbose --config=devspace.yaml    # start standalone sync for detailed output
```
This will show all files and errors directly and will only provide the sync output which makes it much easier to track down issues.

Thanks again to @jessebye for reporting this in the first place and to @abh for providing a reproducible example project.

**Comment by abh** [comment]
I can reproduce it without anything changing (and without the .git) repository.

The key is the big file; without that you get the appropriate permission denied error on writing files.

I am guessing the correlation with .git might be a bunch of files marked read-only.

**Comment by LukasGentele** [comment]
Ok... then I am not able to reproduce this. As soon as I add the `.git` folder to excludePaths, everything works fine. The big file is being uploaded into the container as well. No EOF issue for me anymore. 

But even when you add `.git/` to `excludePaths` and **not** just to `upload/downloadExcludePaths`, you are still experiencing the EOF issue?

If yes, please deploy the project and send me the output of this command:
```
devspace sync --verbose --config=devspace.yaml
```

**Comment by abh** [comment]
oh, okay -- this is confusing, sorry!  Because I added the test.bin file to the repository then the repository .pack files are big too which makes the test look the same without and with the big file.

Kaniko also seems to (sometimes?) add the /app directory with improbable 777 permissions; I fixed that.

I think the error ("we can't sync the files, you did something wrong") is legitimate; the bug (as I see it) is the missing error message if the amount of data to upload is too big.

I force pushed a new version; either do a fresh clone or do this to clean up the .git repository size (it should just be 1 or 200k).

```
git remote update
git reset --hard origin/master
git reflog expire --expire-unreachable=all --all --expire=all
git gc --prune=now --aggressive
```

then run (as you know) `devspace dev`

In this case I get this (helpful!) error in the sync log:

```
 info    Upstream - Retry upload because of error: rpc error: code = Unknown desc = untar all: untarNext: create /app/.git/objects/pack/pack-1ad94d38bce3bce123782400dc37a91299ad4dd5.idx: open /app/.git/objects/pack/pack-1ad94d38bce3bce123782400dc37a91299ad4dd5.idx: permission denied   2020-04-10T16:56:44-07:00
```

(because of the 0444 permissions on those files and however the files are written). It says the upload / initial sync completed all the same.

Now run `devspace run testfiles`. It will sync this file file (because uploading 10MB works fine, it works with a much bigger file, too).
 
Okay, leave the big file in place and do a new deployment (`devspace dev -d`) and it will error out without a helpful error:

```
 info    Upstream - Retry upload because of error: EOF         2020-04-10T20:01:29-07:00
```

Without having looked at the code by guess is that the client gets disconnected on the error while uploading and doesn't get to read the error message. When there's less to upload this doesn't happen. To fix it you'd need to have the server (pod) side keep the connection open and let the client disconnect after reading the error; or some variation of that.

**Comment by LukasGentele** [comment]
Ok great! Sorry I misunderstood at first. Thanks for clarifying this. I talked to @FabianKramm and discussed this with him. We can actually fix the error message and make it more helpful. We will also add a suggestion on how to fix it in the fatal message which will suggest to add the file to excludePaths. Additionally, I'll do what I wrote in the comment above regarding init and docs. Does that sounds good? Did I miss anything else?

PS: File permissions in kaniko can get messed up when `snapshotMode: time` (which is our default because of the faster builds). And you are right, this really just happens sometimes because as far as I understood it when playing with it, it depends on if previous layers have been cached. I saw that you already fixed this by setting `snapshotMode: full` in your example. This is the correct way to fix this. Unfortunately, kaniko project seems not to be fixing this. It has been an issue for a while and their README states "Note that these issues are currently theoretical only." although this seems odd because we had many users run into this issue with applications where permissions matter for execution (e.g. PHP): https://github.com/GoogleContainerTools/kaniko#mtime-and-snapshotting

**Comment by jessebye** [comment]
We have `.git` excluded and are trying to sync only about 10M worth of files, but still experience this issue. The logs look something like this: 

```
[info]   Upstream - Upload 1566 create changes (size 6542410)
[error]  Sync Error on /Users/jessebye/repos/trading-api/modules/service/target/classes: upstream: apply changes: apply creates: upload archive: upload send: EOF
[info]   Sync stopped
[info]   Restarting sync...
```

**Comment by abh** [comment]
@jessebye Try moving the big files out of place and see if the rest works.  There's likely a permission error making the sync not work. The bug, as I have seen it, is not a problem syncing when it should have worked but "just" poor error reporting.

**Comment by jessebye** [comment]
@abh ah, thanks. I had thought 10M was not very big, but I tried a folder with 8KB of files and saw the actual error was permissions-related. We are using `jib` to build the image, and the files in the container are owned by root. I suppose that will be a problem for devspace sync...

```
[info]   Upstream - Retry upload because of error: rpc error: code = Unknown desc = untar all: untarNext: create /app/classes/state_machines/deploy/connected_trading/main.json: open /app/classes/state_machines/deploy/connected_trading/main.json: permission denied
```

**Comment by abh** [comment]
What I did for my use case is for development I use a different Dockerfile that makes the development environment files owned by the user the application is running as …

An alternative might be starting your app with su-exec or similar and letting the devspace tools run as root, I didn’t try that.

**Comment by jessebye** [comment]
That seems reasonable. I do wonder if `skaffold` is able to do syncing with jib, and how it handles it?

**Comment by FabianKramm** [comment]
@abh thanks so much for detailed description and error repository! I fixed the issue with the error message (you were correct the error message was not transferred anymore because the transport was closed already) and from the next devspace version onwards, always the correct error message will be displayed regardless of the size of the files.  

@jessebye skaffold is doing a similar approach as we do regarding the sync, they basically run a `kubectl cp` ([Code](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/sync/kubectl.go#L49)) for each file in the background with the same permissions and user as we do and if it does not work with devspace, chances are very good it won't work with skaffold either.  

So I guess it will come down that you have to change the user the sync is running with and you have two options here I think:
1. The easiest and probably the best in your case is to change the active user of the pod, you can configure that in kubernetes like this
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
spec:
  securityContext:
    runAsUser: 0
    runAsGroup: 0
  ...
```
2. In the case your pod does not need the code directly for the entrypoint, you can add a sidecar container with an empty dir that is root where you can sync to and mount the empty dir to both containers:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
spec:
  containers:
  - name: container1
     ...
     volumeMounts:
     - name: code
        mountPath: /mycodedirectory
  - name: newsidecarcontainer
     image: ubuntu
     command: ["sleep"]
     args: ["infinity"]
     volumeMounts:
     - name: code
        mountPath: /code
  volumes:
  - name: code
     emptyDir: {}
```
and in your devspace.yaml you define the new container as sync target:
```yaml
version: v1beta9
images:
  ...
deployments:
  ...
dev:
  sync:
  - imageName: default
     containerName: newsidecarcontainer
     containerPath: /code
```

**Comment by jessebye** [comment]
@FabianKramm thanks for the tips, we got it to work by setting the user/group to `0`! 🎉

**Comment by LukasGentele** [comment]
Looks like we got both specific cases solved, the error message has been improved in v4.12 and the docs contain a [sync troubleshooting guide](https://devspace.sh/cli/docs/guides/file-synchronization#troubleshooting) now. Closing this for now. Thanks again everyone for helping to narrow these issues down!
